File: src/main/java/org/hydev/mcpm/client/arguments/ArgsParser.java
package org.hydev.mcpm.client.arguments;

import net.sourceforge.argparse4j.ArgumentParsers;
import net.sourceforge.argparse4j.inf.ArgumentParser;
import net.sourceforge.argparse4j.inf.ArgumentParserException;
import org.hydev.mcpm.client.arguments.parsers.CommandParser;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.List;
import java.util.function.Consumer;

/**
 * Handles parsing command arguments into CommandEntry objects (to be dispatched to Controller).
 * For steps to add a new command, see Command.java.
 *
 * @author Azalea (https://github.com/hykilpikonna)
 * @since 2022-10-30
 */
public class ArgsParser
{
    private final ArgumentParser parser;
    private final Consumer<String> log;

    /**
     * Creates a new ArgsParser object with a list of CommandParsers.
     *
     * @param allParsers A list of all CommandParser objects that this ArgsParser can parse.
     */
    public ArgsParser(List<CommandParser> allParsers, Consumer<String> log) {
        this.log = log;

        parser = ArgumentParsers
            .newFor("loader")
            .build()
            .defaultHelp(true);

        var parsers = parser.addSubparsers();

        for (var parser : allParsers) {
            var subparser = parser.configure(parsers);

            if (subparser != null) {
                subparser
                    .setDefault("handler", parser);
            }
        }
    }

    public void parse(String[] arguments) throws ArgumentParserException {
        parse(arguments, log);
    }

    /**
     * Parses arguments into a CommandEntry object.
     *
     * @param arguments A list of string arguments.
     *                  Example: [ "load", "pluginA", "pluginB" ].
     * @param log Logger
     * @throws ArgumentParserException Thrown when arguments are not parsed correctly.
     *                                 For default handling, pass this to ArgsParser#fail.
     */
    public void parse(String[] arguments, Consumer<String> log) throws ArgumentParserException {
        var namespace = parser.parseArgs(arguments);

        Object handleObject = namespace.get("handler");

        if (!(handleObject instanceof CommandParser handler)) {
            throw new ArgumentParserException("Unrecognized command.", parser);
        }

        handler.run(namespace, log);
    }

    /**
     * Writes the error message to stdout (along with help details if needed).
     *
     * @param e The error object that was caught from #parse.
     */
    public void fail(ArgumentParserException e) {
        StringWriter writer = new StringWriter();
        PrintWriter printer = new PrintWriter(writer);

        parser.handleError(e, printer);

        log.accept(writer.toString());
    }

    /**
     * Default help string that can be presented to the user if help dialog is needed.
     *
     * @return A string containing help information on each command.
     */
    public String help() {
        return parser.formatHelp();
    }

    /**
     * ArgsParser demo main.
     *
     * @param args Arguments are ignored.
     */
    public static void main(String[] args) {
        var parser = CommandsFactory.baseArgsParser();

        try {
            parser.parse(args);
        } catch (ArgumentParserException e) {
            parser.fail(e);
        }
    }
}

File: src/main/java/org/hydev/mcpm/client/arguments/ArgsParser.kt
package org.hydev.mcpm.client.arguments

import net.sourceforge.argparse4j.ArgumentParsers
import net.sourceforge.argparse4j.inf.*
import org.hydev.mcpm.client.arguments.parsers.CommandHandler
import org.hydev.mcpm.client.arguments.parsers.CommandParser
import org.hydev.mcpm.client.interaction.ILogger
import java.io.PrintWriter
import java.io.StringWriter

/**
 * Handles parsing command arguments into CommandEntry objects (to be dispatched to Controller).
 * For steps to add a new command, see Command.java.
 */
class ArgsParser(val rawSubparsers: List<CommandParser>)
{
    val help: String
    val parser: ArgumentParser = ArgumentParsers.newFor("mcpm").addHelp(false).build()

    /**
     * Creates a new ArgsParser object with a list of CommandParsers.
     *
     * @param allParsers A list of all CommandParser objects that this ArgsParser can parse.
     */
    init
    {
        val parsers = parser.addSubparsers()
        for (parser in rawSubparsers)
        {
            val subparser = parsers.addParser(parser.name, false)
            parser.configure(subparser)
            subparser.setDefault("handler", parser)
            subparser.addArgument("-h", "--help").action(PrintHelpAction(subparser))
        }

        // Create help string and help command
        help = "mcpm: Minecraft Plugin Package Manager\n" +
            rawSubparsers.filter { it.description.isNotBlank() }
                .map { "&f/mcpm ${it.name} &6- ${it.description}&r\n" }.joinToString("") +
            "To view the help message of a command, use /mcpm <command> -h"

        val helpSub = parsers.addParser("help", false)
        helpSub.setDefault("handler", object : CommandHandler
        {
            override suspend fun run(details: Namespace, log: ILogger) = log.print(help)
        })
    }

    /**
     * Parses arguments into a CommandEntry object.
     *
     * @param arguments A list of string arguments. Example: [ "load", "pluginA", "pluginB" ].
     * @param log Logger
     * @throws ArgumentParserException Thrown when arguments are not parsed correctly.
     * For default handling, pass this to ArgsParser#fail.
     */
    @Throws(ArgumentParserException::class)
    suspend fun parse(arguments: Array<String>, log: ILogger)
    {
        // If no args are present, add help
        if (arguments.isEmpty())
        {
            log.print(help)
            return
        }

        try
        {
            val namespace = parser.parseArgs(arguments)
            val handleObject = namespace.get<CommandHandler>("handler")
                ?: throw ArgumentParserException("Unrecognized command.", parser)
            handleObject.run(namespace, log)
        }
        catch (e: HelpException)
        {
            log.print("&e" + e.help())
        }
    }

    /**
     * Writes the error message to stdout (along with help details if needed).
     *
     * @param e The error object that was caught from #parse.
     */
    fun fail(e: ArgumentParserException?, log: ILogger)
    {
        val writer = StringWriter()
        val printer = PrintWriter(writer)
        parser.handleError(e, printer)
        log.print("&c$writer")
    }
}
